### ISSUE TITLE
Upload artifacts from a pull request

### ISSUE THREAD

**Issue by sjackman** [travis-ci/travis-ci#5579]

**Comment by mikemcquaid** [comment]
Homebrew would also really love this feature; it's the single biggest blocker to us adopting Travis CI more widely.

**Comment by rodneyrehm** [comment]
While this feature request is focused on artifact upload, I think the issue lies deeper. It seems like a lot of people would like some sort of way around the [pull request security restrictions](https://docs.travis-ci.com/user/pull-requests#Security-Restrictions-when-testing-Pull-Requests). I'd like to see a combination of [this suggestions](https://github.com/travis-ci/travis-ci/issues/1946#issuecomment-55500556) and the herein proposed source modification detection for enforcing security.

The number of places one could inject code to extract the decrypted env is fairly limited. It's only .travis.yml and code that is actually *executed* on Travis. A glob should be able to find those files.

If the project's setting "encrypted env in PR" is activated and the PR does not touch any of the files identified by the project's setting "restricted files for encrypted env in PR", secure environment variables can be decrypted and exposed to the build scripts. 

This way each project can choose how they want/need to handle their security.

Without touching this policy, Travis users are either forced to awkwardly dance around PRs - or worse decide that "security" is not worth the hassle, as Angular [seems to have done](https://github.com/angular/angular.js/blob/79b6d557922bfe70863b5012100d02f33b940bed/.travis.yml#L25-L26). People even got "creative" with this as QUnit [demonstrates](https://github.com/jquery/qunit/blob/master/.travis.yml#L7-L8).

**Comment by sjackman** [comment]
> The number of places one could inject code to extract the decrypted env is fairly limited. It's only .travis.yml and code that is actually executed on Travis. A glob should be able to find those files.

If the project runs a unit test suite with good code coverage, then all of the project's code is run on Travis, by design.

**Comment by rodneyrehm** [comment]
> If the project runs a unit test suite with good code coverage, then all of the project's code is run on Travis, by design.

fair point. but not every project runs code that has access to the environment. it depends on the project, that's why this should be configurable.

**Comment by sjackman** [comment]
How likely is it that this feature request will be implemented in the next month? I'd like to know whether I should start working on a plan B.

**Comment by sjackman** [comment]
@BanzaiMan Hi, Hiro. What's the status of this feature request? Any hope for its being implemented?

**Comment by BanzaiMan** [comment]
@sjackman Not any time soon. Sorry.

**Comment by BanzaiMan** [comment]
@sjackman This needs some careful considerations, and that'll take time. Sorry.

**Comment by sjackman** [comment]
I understand. I was only hoping for a status update. Thanks for that.

**Comment by shoogle** [comment]
We'd also like this feature for MuseScore so that we can deploy binaries from pull requests. Ideally the secure variables would be set for a subsection of commands in .travis.yml (e.g. in the "after_script" section) and there would be a list of secure files that mustn't be modified in a PR. If a secure file was modified then the build would proceed without setting the secure variables.

**Comment by sjackman** [comment]
Peter's suggestion would work well for Linuxbrew as well. We have the same use case, deploying binaries from pull requests.

**Comment by shoogle** [comment]
Security can be guaranteed by having the option to only set some (but not necessarily all) environment variables on pull requests. That way we could use different credentials for deploying pull requests as opposed to our official nightly builds, and the official credentials would never be exposed to unvetted code. In a worst-case scenario, a malicious pull request might somehow expose the key for the account hosting the pull request builds, but the security of the official builds could never be compromised by a PR.

**Comment by 130s** [comment]
Just FYI. Since I just found this ticket while I just started testing upload from PR today, I might be misunderstanding something. But I was able to upload the artifact to my S3, basically following [this blog post](https://blog.travis-ci.com/2014-09-09-improved-artifacts-like-dinosaurs-but-cooler/) that was published on 2014. I set the security credentials at `https://travis-ci.org/%ORG%/%REPO%/settings` as suggested in the blog.

Some things look strange though:
- [Log from Travis says `Artifacts support disabled for pull requests`](https://travis-ci.org/130s/user_common_travis/builds/121880934#L3130) (but I see artifacts uploaded).
- On S3, instead of the specific file that I created during my Travis run, it seems all files in my repository are uploaded in addition to the artifact file.
  - I thought `artifacts` addon use the output from `git ls-files -o` to find the artifacts by default, and the command returns nothing on my local repo (so I expect nothing other than artifact file to be uploaded).
  - I guess this is annoyingly better than nothing to me.

If it's helpful, I'm working on this at https://github.com/130s/user_common_travis/pull/7

**Comment by shoogle** [comment]
@130s, I think it worked for you because you submitted your PR into your own repo, not someone else's repo, so the PR was considered "trustworthy".

Now imagine made a PR for your repo with this in .travis.yml:
```yaml
script:
  echo \"key: $ARTIFACTS_KEY\"
  echo \"secret: $ARTIFACTS_SECRET\"
```
If Travis also considered my PR trustworthy then I could just read your secure credentials from the build log and then hack your S3, or even just keep submitting PRs and uploading to your S3 until it got full, probably costing you lots of money in the process!

Hence, Travis considers all external PRs to be "untrustworthy" and doesn't set the secure variables. But sometimes it would be useful to set secure variables on a PR despite the risks, and that's what this ticket is about. The idea is you would declare that PRs are not allowed to change .travis.yml so your credential stay safe, and you could use a free storage service to avoid paying for spam uploads.

**Comment by ericfont** [comment]
I notice GitLab supports storing artifacts from merge requests: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/1584

Maybe something can be learned from that?

**Comment by Christian-Seematter** [comment]
I think this is still a very valuable feature. It is especially relevant for pull requests to have build artifacts when a build fails, so the problem can be resolved before going into master.

My team currently has issues that occur travis only, so debugging that outside of any deployed branches is essential.

**Comment by elopio** [comment]
Fwiw, we are currently uploading artifacts using transfer.sh. Something like this:
```
sudo snap install transfer
transfer path/to/the/file
```

**Comment by sjackman** [comment]
I would still like to be able to upload artifacts from a fork pull request. As far as I know, it's not possible using TravisCI. I use CircleCI instead, where artifacts of fork pull requests are transferred to AWS S3. I then use a webhook to transfer the artifacts from S3 to Bintray using AWS Lambda.

**Comment by artigat1** [comment]
If you're closing it, does that mean artifacts can now be uploaded to S3 from a PR? If so, then how? I have a single e2e test that fails only when running in Travis. I don't want to merge the PR and break the dev build, but I'd like the artifacts so I can track down what is going wrong.

**Comment by Fandekasp** [comment]
It seems like security issues are mostly related to PRs from forked projects. Therefore, would it be possible to enable artifacts upload for PRs created from the same repository, or to map the PR assignee with the list of members from the project teams & collaborators ?

**Comment by haraldF** [comment]
Some suggestions to bypass the security issues:

- Allow variable injection only in specific "secure" jobs
- These "secure" jobs could be encrypted with the travis CI key, e.g. the person pushing the pull-request will only see some binary blob at the end of the .travis.yml

or

- Pre-define some jobs in the Travis UI, so only the owner of the repository will see that job. Only those jobs will have access to secure variables
- Make that job callable from .travis.yml

or

- Do a multi-stage build, inject secure variables only in the second stage
- Second stage must come from a yml file checked into the master branch, cannot be modified / spied upon by the PR author

**Comment by sjackman** [comment]
CircleCI also uploads artifacts to their own servers (stored on Amazon S3).

**Comment by ThomasF34** [comment]
Hey ! :wave: What's the status of this feature-request ?

**Comment by shoogle** [comment]
Meanwhile, GitHub Actions has this feature.

**Comment by sjackman** [comment]
👍 https://github.com/actions/upload-artifact
